core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -139,7 +140,13 @@ public String defaultHostname() {
      * @see Server#activePort()
      */
     public Map<InetSocketAddress, ServerPort> activePorts() {
-        return unmodifiableActivePorts;
+        if (isMutatingActivePorts) {
+            synchronized (activePorts) {
+                return Collections.unmodifiableMap(new LinkedHashMap<>(activePorts));
+            }
+        } else {
+            return unmodifiableActivePorts;
```

Comment:
I think we should never give this without making a copy. Consider the following case:

1. A user keeps the reference to `unmodifiableActivePorts` while `Server` is not mutating anything.
2. A user starts the `Server`.
3. A user tries to access the `unmodifiableActivePorts` retrieved at (1) while the `Server` mutates `activePorts`.

Reply:
Ah, I see. You're right. There's no way to get out of copying. :-(